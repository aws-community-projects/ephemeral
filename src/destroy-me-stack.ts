import { StackProps, Duration, Stack } from 'aws-cdk-lib';
import { DestroyMeConstruct } from './destroy-me';
import { Construct } from 'constructs';

export interface DestroyMeStackProps extends StackProps {
  readonly destroyMeDuration?: Duration
  readonly destroyMeEnable?: boolean
}
export class DestroyMeStack extends Stack {
  constructor (
    scope: Construct,
    id: string,
    props: DestroyMeStackProps = {
    } as DestroyMeStackProps,
  ) {
    super(scope, id, props);
    const {
      destroyMeEnable: enable,
      destroyMeDuration: duration,
    } = props;

    if (enable) {
      new DestroyMeConstruct(this, 'DestroyMeConstruct', {
        duration: duration ?? Duration.days(7),
      });
    } else {
      console.log('DestroyMe: disabled');
    }
  }
}
