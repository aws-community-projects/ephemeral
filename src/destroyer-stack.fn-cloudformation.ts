import type { EventBridgeEvent } from 'aws-lambda';
import { CloudFormationClient,
  DescribeStacksCommand } from '@aws-sdk/client-cloudformation';
import { DynamoDBClient } from '@aws-sdk/client-dynamodb'; // ES6 import
import { DynamoDBDocumentClient, PutCommand } from '@aws-sdk/lib-dynamodb';

const marshallOptions = {
  // Whether to automatically convert empty strings, blobs, and sets to `null`.
  convertEmptyValues: false, // false, by default.
  // Whether to remove undefined values while marshalling.
  removeUndefinedValues: true, // false, by default.
  // Whether to convert typeof object to map attribute.
  convertClassInstanceToMap: true, // false, by default.
};

const unmarshallOptions = {
  // Whether to return numbers as a string instead of converting them to native JavaScript numbers.
  wrapNumbers: false, // false, by default.
};

const translateConfig = {
  marshallOptions,
  unmarshallOptions,
};

const client = new DynamoDBClient({
});
const ddbDocClient = DynamoDBDocumentClient.from(client, translateConfig);
const cf = new CloudFormationClient({
});

export const handler = async (
  event: EventBridgeEvent<string, any>,
): Promise<void> => {
  if (!process.env.DESTROY_TABLE_NAME) {
    return;
  }
  const StackName = event.detail['stack-id'];
  const describeCommand = new DescribeStacksCommand({
    StackName,
  });
  const stacks = await cf.send(describeCommand);

  const stack = stacks.Stacks?.[0];

  const stackLife = stack?.Tags?.find((tag) => tag.Key === 'STACK_LIFE')?.Value;
  if (stackLife) {
    try {
      await ddbDocClient.send(
        new PutCommand({
          TableName: process.env.DESTROY_TABLE_NAME,
          Item: {
            pk: stack.StackName,
            ttl: Math.ceil(new Date().getTime() / 1000 + Number(stackLife)),
          },
        }),
      );
    } catch (e) {
      console.log(e);
    }
  }
};
