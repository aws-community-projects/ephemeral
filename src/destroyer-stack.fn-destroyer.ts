import type { DynamoDBStreamEvent } from 'aws-lambda';
import { unmarshall } from '@aws-sdk/util-dynamodb';
import { AttributeValue } from '@aws-sdk/client-dynamodb';
import { CloudFormationClient, DeleteStackCommand } from '@aws-sdk/client-cloudformation';

export const handler = async (event: DynamoDBStreamEvent): Promise<void> => {
  const stacksToDestroy = event.Records.reduce<string[]>((p, record) => {
    if (record.eventName === 'REMOVE') {
      if (record.dynamodb?.OldImage) {
        const oldImage = record.dynamodb.OldImage as Record<
        string,
        AttributeValue
        >;
        const item = unmarshall(oldImage);
        const currentTimeInSeconds = new Date().getTime() / 1000;
        if (item.ttl > currentTimeInSeconds) {
          // item was manually removed and not expired
          console.log('item was manually removed and not expired', currentTimeInSeconds, item.ttl);
          return [...p];
        }
        return [...p, item.pk];
      }
    }
    return [...p];
  }, []);
  console.log(stacksToDestroy);
  if (stacksToDestroy.length > 0) {
    const client = new CloudFormationClient({
    });
    await Promise.all(
      stacksToDestroy.map(
        async (stackName) =>
          await client.send(
            new DeleteStackCommand({
              StackName: stackName,
            }),
          ),
      ),
    );
  }
};
