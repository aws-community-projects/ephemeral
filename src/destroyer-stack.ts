import { Duration, RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib';
import { AttributeType,
  BillingMode,
  StreamViewType,
  Table } from 'aws-cdk-lib/aws-dynamodb';
import { Construct } from 'constructs';
import { Runtime, StartingPosition } from 'aws-cdk-lib/aws-lambda';
import { DynamoEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { PolicyStatement, Effect } from 'aws-cdk-lib/aws-iam';
import { Rule } from 'aws-cdk-lib/aws-events';
import { LambdaFunction, SnsTopic } from 'aws-cdk-lib/aws-events-targets';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { existsSync } from 'fs';
import { join } from 'path';

export class DestroyerStack extends Stack {
  constructor (scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    const tableName = 'destroyer';
    const table = new Table(this, tableName, {
      tableName,
      partitionKey: {
        name: 'pk',
        type: AttributeType.STRING,
      },
      billingMode: BillingMode.PAY_PER_REQUEST,
      removalPolicy: RemovalPolicy.DESTROY,
      timeToLiveAttribute: 'ttl',
      stream: StreamViewType.NEW_AND_OLD_IMAGES,
    });

    const local = existsSync(join(__dirname, 'destroyer-stack.fn-cloudformation.ts'));
    const cloudformationFn = new NodejsFunction(this, 'fn-cloudformation', {
      runtime: Runtime.NODEJS_18_X,
      memorySize: 1024,
      timeout: Duration.minutes(5),
      entry: join(__dirname, local ? 'destroyer-stack.fn-cloudformation.ts' : 'destroyer-stack.fn-cloudformation.js'),
      initialPolicy: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: [
            'cloudformation:Describe*',
            'cloudformation:Get*',
            'cloudformation:List*',
          ],
          resources: ['*'],
        }),
      ],
    });
    table.grantReadWriteData(cloudformationFn);
    cloudformationFn.addEnvironment('DESTROY_TABLE_NAME', table.tableName);
    new Rule(this, 'cloudformation-rule', {
      eventPattern: {
        source: ['aws.cloudformation'],
        detailType: ['CloudFormation Stack Status Change'],
      },
      targets: [new LambdaFunction(cloudformationFn)],
    });

    const destroyFn = new NodejsFunction(this, 'fn-destroyer', {
      runtime: Runtime.NODEJS_18_X,
      memorySize: 1024,
      timeout: Duration.minutes(5),
      entry: join(__dirname, local ? 'destroyer-stack.fn-destroyer.ts' : 'destroyer-stack.fn-destroyer.js'),
    });

    destroyFn.addEventSource(
      new DynamoEventSource(table, {
        startingPosition: StartingPosition.LATEST,
      }),
    );
    destroyFn.addToRolePolicy(
      new PolicyStatement({
        actions: ['cloudformation:DeleteStack'],
        resources: ['*'],
        effect: Effect.ALLOW,
      }),
    );

    const failTopic = new Topic(this, 'fail-topic');
    new Rule(this, 'delete-failed-rule', {
      eventPattern: {
        source: ['aws.cloudformation'],
        detailType: ['CloudFormation Stack Status Change'],
        detail: {
          resourceStatus: ['DELETE_FAILED'],
        },
      },
      targets: [new SnsTopic(failTopic)],
    });
  }
}
