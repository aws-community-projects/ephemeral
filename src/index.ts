export * from './self-destruct-aspect';
export * from './self-destruct-construct';
export * from './self-destruct-stack';
export * from './make-destroyable-aspect';
export * from './destroy-me-construct';
export * from './destroy-me-stack';
export * from './destroyer-stack';
