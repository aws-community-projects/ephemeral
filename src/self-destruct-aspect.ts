import { MakeDestroyableAspect as SelfDestructAspect } from './make-destroyable-aspect';
// backwards compatible
export { SelfDestructAspect };
