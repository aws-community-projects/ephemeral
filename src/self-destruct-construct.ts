import { CallAwsService } from 'aws-cdk-lib/aws-stepfunctions-tasks';
import { Choice,
  Condition,
  Map,
  Pass,
  StateMachine,
  Succeed,
  Wait,
  WaitTime } from 'aws-cdk-lib/aws-stepfunctions';
import { Construct } from 'constructs';
import { Duration, Stack } from 'aws-cdk-lib';
import { AwsCustomResource,
  AwsCustomResourcePolicy,
  PhysicalResourceId } from 'aws-cdk-lib/custom-resources';

export interface SelfDestructConstructProps {
  readonly duration: Duration
}

export class SelfDestructConstruct extends Construct {
  constructor (scope: Construct, id: string, props: SelfDestructConstructProps = {
  } as SelfDestructConstructProps) {
    super(scope, id);

    const { duration } = props;

    const listExecutions = new CallAwsService(this, 'ListExecutions', {
      action: 'listExecutions',
      iamAction: 'states:ListExecutions',
      iamResources: ['*'],
      parameters: {
        'StateMachineArn.$': '$$.StateMachine.Id',
        StatusFilter: 'RUNNING',
      },
      service: 'sfn',
    });

    const executionsMap = new Map(this, 'ExecutionsMap', {
      inputPath: '$.Executions',
    });

    const stopExecution = new CallAwsService(this, 'StopExecution', {
      action: 'stopExecution',
      iamAction: 'states:StopExecution',
      iamResources: ['*'],
      parameters: {
        Cause: 'Superceded',
        'ExecutionArn.$': '$.ExecutionArn',
      },
      service: 'sfn',
    });

    executionsMap.iterator(
      new Choice(this, 'NotSelf?')
        .when(
          Condition.not(
            Condition.stringEqualsJsonPath('$.ExecutionArn', '$$.Execution.Id'),
          ),
          stopExecution,
        )
        .otherwise(new Pass(this, 'self')),
    );

    const wait = new Wait(this, 'Wait', {
      time: WaitTime.duration(duration),
    });
    const wasDelete = new Choice(this, 'WasDelete?')
      .when(
        Condition.stringEquals('$$.Execution.Input.Action', 'Delete'),
        new Succeed(this, 'DeleteSuccess'),
      )
      .otherwise(wait);

    const deleteStack = new CallAwsService(this, 'DeleteStack', {
      action: 'deleteStack',
      iamAction: 'cloudformation:DeleteStack',
      iamResources: ['*'],
      parameters: {
        'StackName.$': '$$.Execution.Input.StackName',
      },
      service: 'cloudformation',
    });

    const finished = new Succeed(this, 'Finished');

    listExecutions.next(executionsMap);
    executionsMap.next(wasDelete);
    wait.next(deleteStack);
    deleteStack.next(finished);

    const sm = new StateMachine(this, 'SelfDestructMachine', {
      definition: listExecutions,
    });

    new AwsCustomResource(this, 'SelfDestructCR', {
      onCreate: {
        action: 'startExecution',
        parameters: {
          input: JSON.stringify({
            Action: 'Create',
            StackArn: Stack.of(this).stackId,
            StackName: Stack.of(this).stackName,
          }),
          stateMachineArn: sm.stateMachineArn,
        },
        physicalResourceId: PhysicalResourceId.of('SelfDestructCR'),
        service: 'StepFunctions',
      },
      onDelete: {
        action: 'startExecution',
        parameters: {
          input: JSON.stringify({
            Action: 'Delete',
            StackArn: Stack.of(this).stackId,
            StackName: Stack.of(this).stackName,
          }),
          stateMachineArn: sm.stateMachineArn,
        },
        physicalResourceId: PhysicalResourceId.of('SelfDestructCR'),
        service: 'StepFunctions',
      },
      onUpdate: {
        action: 'startExecution',
        parameters: {
          input: JSON.stringify({
            Action: 'Update',
            Version: new Date().getTime().toString(),
            StackArn: Stack.of(this).stackId,
            StackName: Stack.of(this).stackName,
          }),
          stateMachineArn: sm.stateMachineArn,
        },
        physicalResourceId: PhysicalResourceId.of('SelfDestructCR'),
        service: 'StepFunctions',
      },
      policy: AwsCustomResourcePolicy.fromSdkCalls({
        resources: [sm.stateMachineArn],
      }),
    });
  }
}
