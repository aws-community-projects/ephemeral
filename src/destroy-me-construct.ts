import { Aspects, Duration, Stack, Tags } from 'aws-cdk-lib';
import { SelfDestructAspect } from './self-destruct-aspect';
import { Construct } from 'constructs';

export interface DestroyMeConstructProps {
  readonly duration: Duration
}

export class DestroyMeConstruct extends Construct {
  constructor (scope: Construct, id: string, props: DestroyMeConstructProps = {
  } as DestroyMeConstructProps) {
    super(scope, id);

    const { duration } = props;
    Tags.of(Stack.of(this)).add('STACK_LIFE', duration.toSeconds().toString());

    Aspects.of(Stack.of(this)).add(new SelfDestructAspect());
  }
}
