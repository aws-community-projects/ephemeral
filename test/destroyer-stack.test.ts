import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { DestroyerStack } from '../src';

describe('DestroyerStack', () => {
  test('default', () => {
    const app = new App();
    const stack = new DestroyerStack(app, 'MyTestStack');
    const template = Template.fromStack(stack);

    template.resourceCountIs('AWS::Lambda::Function', 2);
    template.resourceCountIs('AWS::DynamoDB::Table', 1);

    expect(template).toMatchSnapshot();
  });
});
