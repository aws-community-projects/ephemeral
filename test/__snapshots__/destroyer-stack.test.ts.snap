// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DestroyerStack default 1`] = `
Object {
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "cloudformationrule02545F98": Object {
      "Properties": Object {
        "EventPattern": Object {
          "detail-type": Array [
            "CloudFormation Stack Status Change",
          ],
          "source": Array [
            "aws.cloudformation",
          ],
        },
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "fncloudformation6A17746B",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "cloudformationruleAllowEventRuleMyTestStackfncloudformationDE0A3E4EA39C70E5": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "fncloudformation6A17746B",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "cloudformationrule02545F98",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "deletefailedrule216BFA59": Object {
      "Properties": Object {
        "EventPattern": Object {
          "detail": Object {
            "resourceStatus": Array [
              "DELETE_FAILED",
            ],
          },
          "detail-type": Array [
            "CloudFormation Stack Status Change",
          ],
          "source": Array [
            "aws.cloudformation",
          ],
        },
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Ref": "failtopic250A40AC",
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "destroyer90CDE342": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AttributeDefinitions": Array [
          Object {
            "AttributeName": "pk",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": Array [
          Object {
            "AttributeName": "pk",
            "KeyType": "HASH",
          },
        ],
        "StreamSpecification": Object {
          "StreamViewType": "NEW_AND_OLD_IMAGES",
        },
        "TableName": "destroyer",
        "TimeToLiveSpecification": Object {
          "AttributeName": "ttl",
          "Enabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Delete",
    },
    "failtopic250A40AC": Object {
      "Type": "AWS::SNS::Topic",
    },
    "failtopicPolicyC77F34EC": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "events.amazonaws.com",
              },
              "Resource": Object {
                "Ref": "failtopic250A40AC",
              },
              "Sid": "0",
            },
          ],
          "Version": "2012-10-17",
        },
        "Topics": Array [
          Object {
            "Ref": "failtopic250A40AC",
          },
        ],
      },
      "Type": "AWS::SNS::TopicPolicy",
    },
    "fncloudformation6A17746B": Object {
      "DependsOn": Array [
        "fncloudformationServiceRoleDefaultPolicy695E4447",
        "fncloudformationServiceRole4D7AB5BB",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "bed8782c811ba28f77568f4a61c0e7766aee2bbe8d3148a01c8031741f54b029.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "DESTROY_TABLE_NAME": Object {
              "Ref": "destroyer90CDE342",
            },
          },
        },
        "Handler": "index.handler",
        "MemorySize": 1024,
        "Role": Object {
          "Fn::GetAtt": Array [
            "fncloudformationServiceRole4D7AB5BB",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "fncloudformationServiceRole4D7AB5BB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "fncloudformationServiceRoleDefaultPolicy695E4447": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "cloudformation:Describe*",
                "cloudformation:Get*",
                "cloudformation:List*",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::GetAtt": Array [
                    "destroyer90CDE342",
                    "Arn",
                  ],
                },
                Object {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "fncloudformationServiceRoleDefaultPolicy695E4447",
        "Roles": Array [
          Object {
            "Ref": "fncloudformationServiceRole4D7AB5BB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "fndestroyerBEE7C6F2": Object {
      "DependsOn": Array [
        "fndestroyerServiceRoleDefaultPolicyD501AE3F",
        "fndestroyerServiceRoleB4DC7288",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "56a73d50bd31b4a9652d5db8481605b4c76c8526583c9d6511e90420f04c6c05.zip",
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "MemorySize": 1024,
        "Role": Object {
          "Fn::GetAtt": Array [
            "fndestroyerServiceRoleB4DC7288",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 300,
      },
      "Type": "AWS::Lambda::Function",
    },
    "fndestroyerDynamoDBEventSourceMyTestStackdestroyer1CD695F6AE037CB6": Object {
      "Properties": Object {
        "BatchSize": 100,
        "EventSourceArn": Object {
          "Fn::GetAtt": Array [
            "destroyer90CDE342",
            "StreamArn",
          ],
        },
        "FunctionName": Object {
          "Ref": "fndestroyerBEE7C6F2",
        },
        "StartingPosition": "LATEST",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "fndestroyerServiceRoleB4DC7288": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "fndestroyerServiceRoleDefaultPolicyD501AE3F": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "dynamodb:ListStreams",
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "dynamodb:DescribeStream",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
              ],
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "destroyer90CDE342",
                  "StreamArn",
                ],
              },
            },
            Object {
              "Action": "cloudformation:DeleteStack",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "fndestroyerServiceRoleDefaultPolicyD501AE3F",
        "Roles": Array [
          Object {
            "Ref": "fndestroyerServiceRoleB4DC7288",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
